packages/custom_lint/lib/src/package_utils.dart

```diff
@@ -115,14 +123,75 @@ Future<PackageConfig?> tryParsePackageConfig(Directory directory) async {
   }
 }
 
+/// Finds the project directory associated with an analysis context root
+///
+/// This is a folder that contains both a `pubspec.yaml` and a `.dart_tool/package_config.json` file.
+/// It is either alongside the analysis_options.yaml file, or in a parent directory.
+Directory findProjectDirectory(
+  Directory directory, {
+  Directory? original,
+  bool missingPackageConfigOkay = false,
+}) {
+  final packageConfigFile = directory.packageConfig;
+  final pubspecFile = directory.pubspec;
+  if (packageConfigFile.existsSync() && pubspecFile.existsSync()) {
+    return directory;
+  }
+  if (pubspecFile.existsSync()) {
+    if (missingPackageConfigOkay) {
+      return directory;
+    }
+    throw PackageConfigNotFoundError._(directory.path);
+  }
+  if (directory.parent.uri == directory.uri) {
+    throw FindProjectError._(original?.path ?? directory.path);
+  }
+  return findProjectDirectory(directory.parent, original: directory);
+}
+
 /// Parse the package config of the given directory.
 ///
 /// Throws if the parsing fails, such as if the file is badly formatted or
 /// does not exists.
 Future<PackageConfig> parsePackageConfig(Directory directory) async {
-  final packageConfigFile = directory.packageConfig;
+  final packageConfigFile = findProjectDirectory(directory).packageConfig;
+
   return PackageConfig.parseBytes(
     await packageConfigFile.readAsBytes(),
     packageConfigFile.uri,
   );
 }
+
+@internal
+
+/// A unified interface for both kinds of file missing errors
+abstract class MissingFileError extends Error {}
```

Comment:
Missing file errors are moved to this file, since we discover them above, I left the parse errors in the other file in case there are errors while reading the files, but the missing files are handled here.